CRM/Civicase/Hook/PostProcess/CaseCategoryPostProcessor.php

```diff
@@ -25,53 +23,19 @@ public function run($formName, CRM_Core_Form &$form) {
       return;
     }
 
-    $caseCategoryMenu = new CaseCategoryMenuService();
-    $caseCategoryCustomData = new CaseCategoryCustomDataType();
-    $caseCategoryCustomFieldExtends = new CaseCategoryCustomFieldExtends();
-
+    $op = $form->getVar('_action');
     $formValues = $form->_submitValues;
-    $formAction = $form->getVar('_action');
-
-    if ($formAction == CRM_Core_Action::DELETE) {
-      $caseCategoryValues = $form->getVar('_values');
-      $caseCategoryMenu->deleteItems($caseCategoryValues['name']);
-      $caseCategoryCustomFieldExtends->delete($caseCategoryValues['name']);
-      $caseCategoryCustomData->delete($caseCategoryValues['name']);
-    }
-
-    if ($formAction == CRM_Core_Action::ADD) {
-      $caseCategoryMenu->createItems($formValues['label']);
-      $caseCategoryCustomFieldExtends->create($formValues['label'], "Case ({$formValues['label']})");
-      $caseCategoryCustomData->create($formValues['label']);
-    }
-
-    if ($formAction == CRM_Core_Action::UPDATE) {
-      $updateParams = [
-        'is_active' => !empty($formValues['is_active']) ? 1 : 0,
-        'icon' => 'crm-i ' . $formValues['icon'],
-      ];
-
-      $caseCategoryMenu->updateItems($form->getVar('_id'), $updateParams);
-    }
-  }
-
-  /**
-   * Returns the option value name given it's id.
-   *
-   * @param int $id
-   *   Option value id.
-   *
-   * @return string
-   *   Option value name.
-   */
-  private function getOptionValueName($id) {
-    $result = civicrm_api3('OptionValue', 'get', [
-      'option_group_id' => 'case_type_categories',
-      'id' => $id,
-      'return' => ['name'],
-    ]);
-
-    return $result['values'][$id]['name'];
+    $caseCategoryValues = $form->getVar('_values');
+    $object = [
+      'id' => $form->getVar('_id'),
+      'name' => $caseCategoryValues['name'],
+      'label' => $formValues['label'],
+      'is_active' => $formValues['is_active'],
+      'icon' => $formValues['icon'],
+    ];
+
+    $handler = new CaseTypeCategoryBaseEventHandler();
+    $handler->run($op, 'OptionValue', $object['id'], $object);
```

Comment:
The handler should not accept the exact parameters the hook accepts. It is a class that should be usable anywhere. As discussed, we need different handlers for create/update/delete, We should not put everything in a run function.